### ISSUE TITLE
Mark slow posts as read

### ISSUE THREAD

**Issue by ai** [hplush/slowreader#153]
We need a way to mark slow posts as read and delete them.

How UI will work:
1. When we open slow post from the list, we mark them `read` but keep in the list
2. User can click `Mark as unread` button to remove `read` state
3. When user change page all read posts will be deleted from database

The problem is that we need 3 states:
1. Unread post
4. Was read recently, but user is on the current page and we should not delete it yet
5. User left page and we can delete it

There are 2 options:
1. Have a separated in-memory storage for read posts. On `Mark as unread` we will add posts again to the log.
2. Have a special `read` marker in Post and some logic to clean all `read` posts from log.

We have a plan for future relevant to thus issue design:
- In the future we plan to have History page with last 50 read post (for the case if you read something but then decide that you need to review it again)

**Comment by pekac** [comment]
Can I take this one?

**Comment by ai** [comment]
Sure!

**Comment by pekac** [comment]
> So my suggestion: mark post as read only after user close the post (change to another post or change post to Fast).

Agreed! We should mark it as read only on a manual user action that means "intentionally navigating away" from the opened post.

> What do you think? Should we add it right now?

Yep 👍

**Comment by pekac** [comment]
Hey @ai !

I'm back on this one. Thought about it and wanted to discuss a bit before I dive in and implement it.

There are 2 things that concern me.

## First one ##

About deleting the opened post - there are way too many actions that can happen once a user opens a post. Some are:
- Navigating away from the page
- Reloading
- Closing the post
- Closing tab
- Closing window
- Leaving the post open and walking away for some IRL action
- etc..

So I would propose the following sequence of actions for "slow" posts once a post is opened:
- push it to a "ready for delete" array
- once a user navigates away from the "slow" page - delete all posts in the array
- if a user clicks "mark as unread" remove it from the delete array

## Second one ##

I'm a bit worried about pagination. Let's say we have 50 posts per page on "slow" and we have 5 pages, total 250 posts.

If go through all 50 posts on the `page=1` and read all of them - after we click on the next page what would happen?
- all 50 posts `ids` would first get pushed to the delete array
- on page change - all 50 posts would get deleted
- now, instead of the total 250 posts we have 200 - and those 200 are posts from 51 to 250, since we deleted the first 50.
- then the offset for the `page=2` would take from 50 to 100 but not of the original 250 posts but from the new 200 posts
- this would mean that we would actually display original `page=3` posts instead of displaying `page=2`, if you got me? we would eat up a whole page of posts due to the wrongly calculated offset.

Even if we patch up pagination to handle offset calculations based on how many we delete, I'm still not sure about the UX where the progress bar will be at the same position (0) and the only thing that will change is the number of segments (pages).

E.g. if we delete 50 as previously mentioned from 250 - we will now still be at page 1 but with total 4 pages.
If we repeat the process and delete another 50, we would still be at page 1 but now, with total 3 pages.

E.g. 2: Even worse than that, if we delete 37 posts, and then click "next page" we would have `page=2` with 50 posts and `page=1` with 13 posts. Where if user goes back to the prev page - he would get 50 posts with some that are from the leftover 13 and some that they just saw on the `page=2`.

## Proposed solution ##

To avoid all of this, my proposed solution would be just to keep all the posts in the delete array, as long as the user is on "slow", no matter which particular sub-page (`page=X`). 
And then, delete all the posts only once a user navigates away from the entire "slow" page. 
Posts that were marked for deletion would have lesser opacity or something that would indicate to the users that it was read.

What do you think about this approach?
Sorry for the long post 🥔

**Comment by ai** [comment]
I am thinking maybe we should use not `toRemove` array but `read` key in the `PostValue`.

Use case:
1. I opened feed and start reading.
2. I read 50 posts, they marked as read on laptop.
3. I got a call and run away on the street (without closing anything on the laptop so read posts was not removed).
4. On the street I open phone and found that those 50 posts are not read.

If we mark them as `read` in the `PostValue` this will sync and phone will render them in gray as read posts. Then phone can even delete them (when user change feed we remove all feed’s post with `read` mark).

**Comment by pekac** [comment]
@ai unassign me please.. it's been too long and I don't think I'll get extra time to work on the app for the time being :(
